### PR TITLE
@validates

### PR DESCRIPTION
**Pull request by philtay**
tests + docs after initial review
#116

### CODE REVIEW

File: marshmallow/schema.py [hunk 1]
```diff
@@ -721,6 +723,21 @@ def __filter_fields(self, field_names, obj, many=False):
...
+
+            try:
+                field = self.fields[validator_kwargs['field']]
+            except KeyError:
+                raise ValueError(
+                    '"{}" field does not exist.'
+                    .format(validator_kwargs['field'])
+                )
+            else:
+                field.validators.append(validator)
```
Comment (sloria) [inline marshmallow/schema.py:739]: Will this result in the same validator being appended multiple times if the Schema is instantiated multiple times?

**Comment by philtay** [comment]
oops... two possible solutions:
1. use the `@validates` decorator to set an attribute on the decorated function and use this attribute to avoid appending the function multiple times.
2. iterate over `cls_fields` in `SchemaMeta.__new__` and append the decorated validators.

which one do you prefer? other alternatives?

**Comment by sloria** [comment]
One possible alternative would be to use `Unmarshaller._run_validator`--as is done for schema validators. Not sure if this is the best solution, but it has an advantage in that it might allow us to validate implicitly-created fields (through the `fields` or `additional` class Meta option). 

This would require a slight modification of `_run_validator` to allow for explicit specification of which field to store an error message on.

**Comment by philtay** [comment]
Uhm, is there any reason for `Field.validators` to be a `list` instead of a `set`? If not, I think using a `set` is the best solution.

**Comment by sloria** [comment]
Field.validators is a list because the order of error messages may be significant in some use cases.

**Comment by philtay** [comment]
Out of curiosity, can you describe a real-world use case where you need or rely on validation order? In any case: currently the invocation order of `@validates_schema` methods is not guaranteed. Have you any objection to the use of an `OrderedSet` instead of a `list` as the `Field.validators` type?

http://code.activestate.com/recipes/576694/

**Comment by sloria** [comment]
The use case I've run into is when you want a deterministic order of error messages displayed to a user, e.g. display more important error messages before less important ones.

`OrderedSet` is already included in the marshmallow source. It could work, but I still think modifying `_run_validator` might be better if only because it would allow validation of implicitly-created (class Meta) fields.

**Comment by philtay** [comment]
Ok, but I don't want to push even more logic in an already semi-spaghettish function. It looks like `_run_validator` is meant for schema-level validation only. I can rename it to `_validate_schema` and add another function `_validate_field`. I will use the latter to implement `@validates`. Ok?

**Comment by sloria** [comment]
Go for it

**Comment by philtay** [comment]
Please find below the method you should review before I open a new and clean PR.

``` python
def _invoke_field_validators(self, data, many):
    for attr_name in self.__processors__[(VALIDATES, False)]:
        validator = getattr(self, attr_name)
        validator_kwargs = validator.__marshmallow_kwargs__[(VALIDATES, False)]
        field_name = validator_kwargs['field']

        try:
            field_obj = self.fields[field_name]
        except KeyError:
            raise ValueError('"{}" field does not exist.'.format(field_name))

        if many:
            for idx, item in enumerate(data):
                try:
                    value = item[field_name]
                except KeyError:
                    pass
                else:
                    self._unmarshal.call_and_store(
                        getter_func=validator,
                        data=value,
                        field_name=field_name,
                        field_obj=field_obj,
                        index=(idx if self.opts.index_errors else None)
                    )
        else:
            try:
                value = data[field_name]
            except KeyError:
                pass
            else:
                self._unmarshal.call_and_store(
                    getter_func=validator,
                    data=value,
                    field_name=field_name,
                    field_obj=field_obj
                )
```

It's a method of the `BaseSchema` class and is invoked in `_do_load` just before `_invoke_validators` in this way:

``` python
self._invoke_field_validators(data=result, many=many)
```

Let me know. If it looks good I'll go ahead with a complete PR.

**Comment by sloria** [comment]
@philtay Yes, that is a reasonable approach. Feel free to proceed

**Pull request closed without merging**